### PR TITLE
More generic value snapping for Slider widgets

### PR DESCRIPTION
**Pull request by ianhi**
## PR Summary
Closes: https://github.com/matplotlib/matplotlib/issues/18562

Adds a `valsnap` argument to the Slider widget to allow specifying the values the slider should snap to. This overrides the `valstep` argument.

I also made a new slider demo to show the use of `valstep` and `valsnap` as I think they would get lost in the current [slider demo](https://matplotlib.org/3.3.1/gallery/widgets/slider_demo.html). For a point of context I have read through the existing tutorial multiple times, but never managed to encode that `valstep` existed until I started working on this PR

- I used `searchsorted` instead of `argmin` as it is significantly faster.
- This works for `idx=0` because then the indexing just becomes `-1` and we compare the largest and smallest elements of the array.
## PR Checklist
I have a few questions here:

1. Documentation:
   a. There isn't any narrative docs on widgets to add this to?
   b. Does this deserve it's own example (perhaps named `slider valstep and valsnap demo`) it feel weird to cram into the current 
2. naming. I chose `valsnap` because it was consistent with the other kwargs but if there are better ideas I'm happy to change it.
3. Should this get both a `whats_new` and a `next_api_changes`?
4. To get the number for `next_api_changes` is the best strategy to look through both open and closed PRs right before opening this?
   - Or I guess do what I've done here and wait until after opening the PR

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/widgets.py
Comment (tacaswell) [inline lib/matplotlib/widgets.py:338]: This isn't a great pattern because now if `valsnap is None`, we silently drop the user's input for `valstep` on the floor.  This is the sort of thing that causes the greatest anger at libraries, you change the thing AND NOTHING HAPPENS.

We should find a way to either fold `valstep` and `valsnap` into one parameter or raise if both are passed.

Folding them into one parameter seems ok (there is a bit of ambiguity else where about plural) and an the code shows, you can generate the snap location from a step.   In other cases where we have colliding kwargs we raise (e.g. norm vs density in hist).

File: lib/matplotlib/widgets.py
Comment (tacaswell) [inline lib/matplotlib/widgets.py:258]: If you are adding a new argument it has to go at the end (or better yet be keyword only!).  This in an API break as if there is someone (admittedly making a poor coding choice) of passing in 13 positional arguments, `valsnap` would get the string intended for `orientation` and likely blow up.

**Review by tacaswell — DISMISSED** [review]
:+1: in principle, but the API needs some tweaking.

**Comment by tacaswell** [comment]
- Just needs a `whats_new` (assuming we un-break the signature change) as no existing code should be broken by it.
- if you are changing the API odds are close to one it will take atleast 1 round of review so open the PR, get number, then add file
- adding narrative docs about widgets would be great (but maybe in a different PR)

File: lib/matplotlib/widgets.py
Comment (ianhi) [reply lib/matplotlib/widgets.py:258]: Yikes. But yes ok, I suppose having millions of users nearly guarantees that that is happening somewhere.

File: lib/matplotlib/widgets.py
Comment (ianhi) [reply lib/matplotlib/widgets.py:338]: Ah indeed. Are you expressing a preference for a single kwarg here? I'm 55/45 towards two kwargs (appropriately ordered) and raising an error if both are passed. I think there is also an opportunity for user anger when they discover that a feature exists but was not very discoverable.

**Comment by ianhi** [comment]
While I'm at it would it be ok to add a kwarg that would allow for eliminating the red line that marks the initial value?
image

Something like `markinit` as a boolean. It could also be nice to be able to change the color of that line so another approach could be to have `init_mark_color` that accepts a color or `None` and if `None` then the line is not drawn?

File: lib/matplotlib/widgets.py
Comment (ianhi) [reply lib/matplotlib/widgets.py:339]: I just pushed a change so it is two kwargs and raises an error if both are given. If need be I'm happy to move to a single `valstep` kwarg

Comment (jklymak) [reply lib/matplotlib/widgets.py:339]: You can just do this above,  before the assignment, as `if valstep is not None and valsnap is not None:...`

Comment (ianhi) [reply lib/matplotlib/widgets.py:342]: I moved the `ValueError` up to be with the other `ValueError`s

**Comment by ianhi** [comment]
I also added another kwarg `initcolor` that can be use to specify the previous hardcoded color of the line marking the initial slider position. It also accepts `None` to not draw the line.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:312]: I'm not a fan of mutually exclusive arguments. Would it be too awkward to accept array-like for *valstep*. (Steps don't need to be equal in size and "You give the step positions explicitly").

Comment (ianhi) [reply lib/matplotlib/widgets.py:312]: I think it would actually simplify the implementation. I had avoided because I tend to find arguments with dual means like that to be confusing, especially if the names seems to clearly preference one interpretation. 

That said, I could've sworn I had a goto example of where I've found this confusing, but I can't actually think of an example. Plus this seems reasonably niche, and with adequate documentation it should still be very discoverable.

I will update it to only accept valstep as either a float or an arraylike.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:298]:
```suggestion
        valstep : float or array-like, default: None
```
(standard convention)

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:307]:
```suggestion
            The color of the line at the *valinit* position. If None then
            no line will be drawn.
```
Can be shorted without loss of information.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:258]:
```suggestion
                 orientation='horizontal', *, initcolor='r', **kwargs):
```
We're nowadays more conservative and add new keyword arguments as keyword-only most of the time.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:352]: I find it a bit awkward to have a public attribute and a private attribute that are decoupled but essentially hold the same information.

I propose this:

```suggestion
        if not isinstance(valstep, Number):
            self.valstep = np.asanyarray(valstep)
            if self.valsnap.ndim =! 1:
                raise ValueError(
                    f"valstep must have 1 dimension but has {self.valstep.ndim}"
                )
            self.valstep = np.sort(self.valstep)
```

We don't need to support passing 0-dim arrays - I think we generally don't throughout the library.

And adapt `_value_in_bounds()` accordingly.

Comment (ianhi) [reply lib/matplotlib/widgets.py:352]: > I find it a bit awkward to have a public attribute and a private attribute that are decoupled but essentially hold the same information.

I was worried that if someone passes in a float for valstep and then later checked the value of `slider.valstep` they would be confused by the fact that it had turned into a numpy array. Or perhaps more worrisome is that that would be backwards incompatible for anyone accessing `slider.valstep` after slider creation. I don't imagine that's many people but I suppose it's non-zero. Does matplotlib have a policy on what counts as backwards incompatible?

Comment (ianhi) [reply lib/matplotlib/widgets.py:352]: > We don't need to support passing 0-dim arrays - I think we generally don't throughout the library.

Ahh, I was doing that to check if it was a float vs an array vs something malformed. Where does `Number` come from? That seems like a better way to test this.

Comment (ianhi) [reply lib/matplotlib/widgets.py:352]: > And adapt _value_in_bounds() accordingly.

Final question. Are you proposing I retain the old slider snapping code in addition to the new snapping to an array code? In your suggestion you eliminated the creation of an array when `valstep` is a float.

Comment (timhoffm) [reply lib/matplotlib/widgets.py:352]: `from numbers import Numbers` (python stdlib)

> Does matplotlib have a policy on what counts as backwards incompatible?

In general, we're very strict with backwards compatibility. Exceptions are case-by-case decisions.

> I was worried that if someone passes in a float for valstep and then later checked the value of `slider.valstep` they would be confused by the fact that it had turned into a numpy array. Or perhaps more worrisome is that that would be backwards incompatible for anyone accessing `slider.valstep` after slider creation.

I was conversely worried that somebody might set `slider.valstep`, but that would not update `_valsnap`. (Though I'm not sure if attribute changes currently work 100% as they do only take effect through `_update()` -> `_value_in_bounds()`.

For now I propose the following pragmatic approach:

- `self.valstep = valstep  # validation deferred to _value_in_bounds()` in `__init__`:
- Move all the conversion logic and checks to `_value_in_bounds`.

This is consistent with handling of all parameters. You pay a slight performance penalty, but that is negligable compared to a draw update.

Comment (ianhi) [reply lib/matplotlib/widgets.py]: Done. Thanks for all the guidance on this by the way! It's very helpful and much appreciated.

Comment (ianhi) [reply lib/matplotlib/widgets.py]: It turns out that `set_val` does not currently call `_value_in_bounds` so I added an optional `kwarg` `validate` to enable `set_val` to cal `value_in_bounds`. Defaults `False` for backwards compat.

Comment (timhoffm) [reply lib/matplotlib/widgets.py]: Thanks for the positive feedback! I'm sorry this is a bit of a hassle, but all the widget stuff seems a bit ad-hoc (that's been before my time :stuck_out_tongue_closed_eyes:).

I think it's an error to allow setting arbitrary values, that do not comply with bounds. In that sense, we don't need a back compat option (as a general rule, we still can fix bugs without back-comaptibility considerations). We should

- always coerce the value via `_value_in_bounds()`
- add an API change note for the changed behavior
- possibly add a warning to `set_val()` if `val` was modified by `value_in_bounds()`.

To keep things more manageable, I propose to defer changing `set_val()` to a separate PR.

**Comment by ianhi** [comment]
> To keep things more manageable, I propose to defer changing set_val() to a separate PR.

:+1: I dropped that commit. Otherwise I can't think of anything else to change here.

File: examples/widgets/slider_snap_demo.py
Comment (timhoffm) [inline examples/widgets/slider_snap_demo.py:25]:
```suggestion
widget_color = 'lightgoldenrodyellow'
```
may be a bit more clear. (Yes the original slider demo uses `axcolor` but that does not mean we cannot improve :smile: ).

File: examples/widgets/slider_snap_demo.py
Comment (timhoffm) [inline examples/widgets/slider_snap_demo.py:30]: Unused. Either use below, or delete. - I tend to go with deletion, the extra indirection makes it a bit harder to follow.

File: examples/widgets/slider_snap_demo.py
Comment (timhoffm) [inline examples/widgets/slider_snap_demo.py:36]: In this place the comment has the same visual level as `# create the sliders` and it is thus a bit surprising that slider creation goes on.

I'd move this  comment below the slider creation as
```suggestion
# initcolor is set to None to remove the line marking the valinit position
```

File: examples/widgets/slider_snap_demo.py
Comment (timhoffm) [inline examples/widgets/slider_snap_demo.py:26]:
```suggestion
ax_freq = plt.axes([0.25, 0.1, 0.65, 0.03], facecolor=axcolor)
```
For readability.

File: examples/widgets/slider_snap_demo.py
Comment (timhoffm) [inline examples/widgets/slider_snap_demo.py:27]:
```suggestion
ax_amp = plt.axes([0.25, 0.15, 0.65, 0.03], facecolor=axcolor)
```

File: examples/widgets/slider_snap_demo.py
Comment (timhoffm) [inline examples/widgets/slider_snap_demo.py:52]:
```suggestion
ax_reset = plt.axes([0.8, 0.025, 0.1, 0.04])
```

File: examples/widgets/slider_snap_demo.py
Comment (timhoffm) [inline examples/widgets/slider_snap_demo.py:38]: `f0` should be a multiple of `valstep`. Otherwise, the slider will implicitly coerce the value.

File: examples/widgets/slider_snap_demo.py
Comment (timhoffm) [inline examples/widgets/slider_snap_demo.py:31]:
```suggestion
allowed_amplitudes = np.concatenate([np.linspace(.1, 5, 100), [6, 7, 8, 9]])
```
otherwise there's an unexpected gap at 6.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:415]:
```suggestion
            valstep = np.asanyarray(self.valstep)
```
Introducing a new name signals that this is something else, but semantically this stays the same concept.

**Review by timhoffm — APPROVED** [review]
Only minor style comments.

File: examples/widgets/slider_snap_demo.py
Comment (ianhi) [reply examples/widgets/slider_snap_demo.py:25]: Or even the more explicit `slider_background_color`?

Comment (ianhi) [reply examples/widgets/slider_snap_demo.py:25]: Also hmm, looking into this I think the slider docstring has an error. In the note at the end it says 

        Additional kwargs are passed on to ``self.poly`` which is the
        `~matplotlib.patches.Rectangle` that draws the slider knob.  See the
        `.Rectangle` documentation for valid property names (``facecolor``,
        ``edgecolor``, ``alpha``, etc.).

but the poly is actually created using `axhspan` which returns a `Polygon` rather than a rectangle https://matplotlib.org/api/_as_gen/matplotlib.pyplot.axhspan.html#matplotlib.pyplot.axhspan

Should that be addressed here or in a follow up?

Comment (ianhi) [reply examples/widgets/slider_snap_demo.py:25]: oh woops. I was a bit confused. The color you highlighted here is passed to plt.axes, but I stand by that docstring being slightly incorrect.

**Comment by ianhi** [comment]
I also added changing the color of the bar for one of the sliders to increase the discoverability of that option.

**Comment by ianhi** [comment]
Does this need another review from @tacaswell given their prior changes requested?

**Comment by timhoffm** [comment]
Every code PR needs 2 positive reviews. Ideally, @tacaswell would do the second review because he has requested changes. But if he doesn't have time any other reviewer is fine. I take the authority to clear the change request if needed.

File: examples/widgets/slider_snap_demo.py
Comment (QuLogic) [inline examples/widgets/slider_snap_demo.py:6]:
```suggestion
You can snap slider values to discrete values using the ``valstep`` argument.
```

File: examples/widgets/slider_snap_demo.py
Comment (QuLogic) [inline examples/widgets/slider_snap_demo.py:44]:
```suggestion
    initcolor=None  # Remove the line marking the valinit position.
)
```

File: lib/matplotlib/widgets.py
Comment (QuLogic) [inline lib/matplotlib/widgets.py:307]: Usually we use `None` to mean default. We also already have `'none'` to mean no colour; do we really need to support `None` here?

File: lib/matplotlib/widgets.py
Comment (QuLogic) [inline lib/matplotlib/widgets.py:421]: I don't understand how this indexing works when you search using a sorted array, and then index the original unsorted one.

File: lib/matplotlib/widgets.py
Comment (ianhi) [reply lib/matplotlib/widgets.py:307]: I find using `None` to mean "don't draw" and colors otherwise to be the most intuitive api. I'd also worry about the discoverability of using `'none'` for example: I'm an advanced user but did not know that you could specify `'none'` as a color. 

> Usually we use None to mean default

I made the default `'r'`  as that's what was hardcoded, but I think it would make more sense for `None` to be the default and specifiying a color otherwise. I don't really see the usecase for drawing a line at the starting position, but I guess this edges into personal preference rather than logic and additionally is maybe backwards incompatible. What about switching the default to `None`?

File: lib/matplotlib/widgets.py
Comment (ianhi) [reply lib/matplotlib/widgets.py:421]: Good catch thank you!

This was a remnant of the original implementation. This is far faster if you presort the array in the `init` method, but now that we're not doing that using `argmin` will be simpler and likely as fast.

Comment (ianhi) [reply lib/matplotlib/widgets.py:421]: It still worked in the all the examples and tests because every example was sorted.

File: lib/matplotlib/widgets.py
Comment (jklymak) [reply lib/matplotlib/widgets.py:307]: `None` means "default" unless we really can't help it.  I you want the default to be `"none"` I guess that could work for you....

Comment (ianhi) [reply lib/matplotlib/widgets.py:307]: Gotcha - thanks.  It probably is the least friction to as `'r'` and not accept `None` so I'll do that + leave a note in the docstring to use `'none'` if you don't want it to be visible.

**Comment by ianhi** [comment]
I add a reference to the new example - and while doing so added the references to the other widget examples except for `menu.py` which doesn't actually use any widgets and i was less sure what to reference.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**
